Clean up Win32Handle implementation

donny-dont opened the pull request:
#### 584193816745c253e3c91c53ac548f70ab5af811
<pre>
Clean up Win32Handle implementation
<a href="https://bugs.webkit.org/show_bug.cgi?id=256498">https://bugs.webkit.org/show_bug.cgi?id=256498</a>

Reviewed by Fujii Hironori.

Make `Win32Handle` look like `MachSendPort` so it has a similar
interface.

* Source/WTF/wtf/PlatformWin.cmake:
* Source/WTF/wtf/win/MemoryFootprintWin.cpp:
* Source/WTF/wtf/win/MemoryPressureHandlerWin.cpp:
* Source/WTF/wtf/win/Win32Handle.cpp: Added.
* Source/WTF/wtf/win/Win32Handle.h:
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.cpp:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/win/ArgumentCodersWin.cpp:
* Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
* Source/WebKit/Platform/IPC/win/IPCSemaphoreWin.cpp:
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
* Source/WebKit/UIProcess/Launcher/win/ProcessLauncherWin.cpp:

Canonical link: <a href="https://commits.webkit.org/263991@main">https://commits.webkit.org/263991@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/cef6e5b8bc6d3630d43c8e636ef421d01516045c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/6306 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/6502 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6681 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7880 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6632 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/6307 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6722 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/6451 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/9524 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/6418 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/6423 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5709 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7950 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3895 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5688 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13573 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/5313 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5759 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5771 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/8016 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/5902 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/6255 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/5116 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/6431 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5666 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/1516 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/1498 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9821 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/6606 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/6036 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/1642 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WTF/wtf/win/MemoryFootprintWin.cpp:
```diff
@@ -44,7 +44,7 @@ size_t memoryFootprint()
     // > The shared data includes pages that contain all instructions your application executes,
     // > including those in your DLLs and the system DLLs. As the working set size increases,
     // > memory demand increases.
-    Win32Handle process(OpenProcess(PROCESS_QUERY_INFORMATION, FALSE, GetCurrentProcessId()));
```
I don't like this line contains the word `Win32Handle` twice. I'd like to write
~~~c++
auto process = Win32Handle::adopt(OpenProcess(PROCESS_QUERY_INFORMATION, FALSE, GetCurrentProcessId()));
~~~
WDYT?

donny-dont replied:
I agree that it’s more verbose than I’d like but I wanted to keep with the UnixFileDescriptor constructs since we have to touch those when working on PlayStation and having a common language is good.

fujii replied:
It's not WebKit's common style. 
For example, GDIObject class has `adoptGDIObject`.
https://github.com/WebKit/WebKit/blob/4c8182f7f87474c644e5cf5eaf01207778b150ff/Source/WTF/wtf/win/GDIObject.h#L105

donny-dont replied:
Ok added an additional convenience method to construct.

kkinnunen-apple commented on Source/WTF/wtf/win/Win32Handle.h:
```diff
@@ -32,13 +32,27 @@ namespace WTF {
-    explicit Win32Handle(HANDLE handle) : m_handle(handle) { }
-    Win32Handle(const Win32Handle& other)
+
+    enum AdoptionTag { Adopt };
```
I think all handle like objects should be eventually have the same interface:
Win32Handle, MachSendRight, UnixFileDescriptor.

For adoption tag: I'd prefer the unfication happens in other direction, e.g. removal of the public adoption constructor and tag.

All webkit types seem to use specific adoption constructor function:
- MachSendRight::adopt
- adoptRef(T*)
- others

This means that there is no other precedent to use adopt tag except UnixFileDescriptor.

Instead, I'd propose we have:
- Win32Handle::adopt
- UnixFileDescriptor::adopt
- MachSendRight::adopt

fujii commented on Source/WTF/wtf/win/Win32Handle.h:
```diff
@@ -24,74 +25,38 @@
-    {
-        *this = WTFMove(other);
-    }
+    using Type = HANDLE;
```
This isn't needed. Please remove it.

fujii commented on Source/WTF/wtf/win/Win32Handle.h:
```diff
@@ -24,74 +25,38 @@
-        *this = WTFMove(other);
-    }
+    using Type = HANDLE;
+    static constexpr Type invalidHandle = INVALID_HANDLE_VALUE;
```
This isn't needed. Please remove it.

fujii commented on Source/WTF/wtf/win/Win32Handle.h:
```diff
@@ -24,74 +25,38 @@
 
-    ~Win32Handle() { clear(); }
+    WTF_EXPORT_PRIVATE static Win32Handle adopt(Type);
+    WTF_EXPORT_PRIVATE static Win32Handle create(Type);
```
This isn't needed. Please remove it.

fujii commented on Source/WTF/wtf/win/Win32Handle.cpp:
```diff
@@ -0,0 +1,91 @@
+    return Win32Handle(handle);
+}
+
+Win32Handle Win32Handle::create(Type handle)
```
This isn't needed. Please remove it. Move this code to `copy()`.

donny-dont committed:
Clean up Win32Handle implementation
https://bugs.webkit.org/show_bug.cgi?id=256498

Reviewed by Fujii Hironori.

Make `Win32Handle` look like `MachSendPort` so it has a similar
interface.

* Source/WTF/wtf/PlatformWin.cmake:
* Source/WTF/wtf/win/MemoryFootprintWin.cpp:
* Source/WTF/wtf/win/MemoryPressureHandlerWin.cpp:
* Source/WTF/wtf/win/Win32Handle.cpp: Added.
* Source/WTF/wtf/win/Win32Handle.h:
* Source/WebCore/platform/graphics/win/MediaPlayerPrivateMediaFoundation.cpp:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/Platform/IPC/win/ArgumentCodersWin.cpp:
* Source/WebKit/Platform/IPC/win/ConnectionWin.cpp:
* Source/WebKit/Platform/IPC/win/IPCSemaphoreWin.cpp:
* Source/WebKit/Platform/win/SharedMemoryWin.cpp:
* Source/WebKit/UIProcess/Launcher/win/ProcessLauncherWin.cpp:

Canonical link: https://commits.webkit.org/263991@main

kkinnunen-apple commented on Source/WTF/wtf/win/Win32Handle.cpp:
```diff
@@ -0,0 +1,106 @@
+        return INVALID_HANDLE_VALUE;
+
+    auto processHandle = ::GetCurrentProcess();
+    HANDLE duplicate;
```
If the call fails, you might have uninitialized value here.. The docs don't say..

fujii replied:
Good point. Filed https://bugs.webkit.org/show_bug.cgi?id=257435
